src/Concordium/Client/Runner.hs

```diff
@@ -1269,6 +1295,61 @@ getCryptographicParametersOrDie bhInput = do
+--  This is used for determining the time at which a cooldown will actually elapse.
+firstPaydayAfter ::
+    -- | Time of the next payday.
+    UTCTime ->
+    -- | Duration of an epoch
+    Types.Duration ->
+    -- | Length of a payday.
+    Types.RewardPeriodLength ->
+    -- | Time at which the cooldown expires.
+    UTCTime ->
+    UTCTime
+firstPaydayAfter nextPayday epochDuration (Types.RewardPeriodLength ep) cooldownEnd =
+    if cooldownEnd <= nextPayday
+        then nextPayday
+        else
+            let timeDiff = Clock.diffUTCTime cooldownEnd nextPayday
+                paydayLength = Types.durationToNominalDiffTime (fromIntegral ep * epochDuration)
+                mult :: Word = ceiling (timeDiff / paydayLength)
+            in  Clock.addUTCTime (fromIntegral mult * paydayLength) nextPayday
+
+-- | Correct a pending change on an account to account for the fact that it will only actually be
+--  released at the following payday.
+correctPendingChange :: BlockHashInput -> Types.AccountInfo -> ClientMonad IO Types.AccountInfo
+correctPendingChange bhi = stakingInfo . pendingChange . effectiveTime $ \time -> do
+    eChainParams <- getResponseValueOrDie =<< getBlockChainParameters bhi
+    case eChainParams of
+        Queries.EChainParametersAndKeys ChainParameters{_cpTimeParameters = SomeParam timeParams} _ -> do
+            let rewardPeriod = timeParams ^. tpRewardPeriodLength
+            rewardStatus <- getResponseValueOrDie =<< getTokenomicsInfo bhi
+            case rewardStatus of
+                Queries.RewardStatusV0{} -> return time
+                Queries.RewardStatusV1{..} -> do
+                    consensusInfo <- getResponseValueOrDie =<< getConsensusInfo
+                    let epochDuration = Queries.csEpochDuration consensusInfo
+                    return $ firstPaydayAfter rsNextPaydayTime epochDuration rewardPeriod time
+        _ -> return time
+  where
+    stakingInfo :: Lens' Types.AccountInfo Types.AccountStakingInfo
+    stakingInfo = lens Types.aiStakingInfo (\x y -> x{Types.aiStakingInfo = y})
+    pendingChange :: Traversal' Types.AccountStakingInfo (Types.StakePendingChange' UTCTime)
+    pendingChange _ Types.AccountStakingNone = pure Types.AccountStakingNone
+    pendingChange f Types.AccountStakingBaker{..} =
+        (\newPendingChange -> Types.AccountStakingBaker{asiPendingChange = newPendingChange, ..})
+            <$> f asiPendingChange
+    pendingChange f Types.AccountStakingDelegated{..} =
+        ( \newPendingChange ->
+            Types.AccountStakingDelegated{asiDelegationPendingChange = newPendingChange, ..}
+        )
+            <$> f asiDelegationPendingChange
+    effectiveTime :: Traversal' (Types.StakePendingChange' t) t
+    effectiveTime _ Types.NoChange = pure Types.NoChange
+    effectiveTime f (Types.ReduceStake amt oldTime) = Types.ReduceStake amt <$> f oldTime
+    effectiveTime f (Types.RemoveStake oldTime) = Types.RemoveStake <$> f oldTime
```

Comment:
This function seems extremely information dense. It would be very helpful to have some comments that explain what is going on, what situation each `case` corresponds to and what the code is doing to address that situation and why.

As it is, I have a very hard time understanding what this function is doing (my Haskell may not be up to snuff in this case but that hopefully doesn't invalidate this point).